spec/jmap/push.mdown

```diff
@@ -1,69 +1,94 @@
+### setPushSubscription
 
-To set the web hook, make a call to *setPushCallback*. It takes the following argument:
+Each session may only have a single push subscription registered. The push subscription is tied to the access token used to create it. Should the access token expire or be revoked, the push subscription MUST be removed by the JMAP server. The client MUST re-register the push subscription after reauthenticating to resume callbacks.
 
-- **callback**: `String|null`
-  The (HTTPS) URL the JMAP server should POST events to. This will replace any previously set URL. Set to `null` to just remove any previously set callback URL.
+To set the push subscription, make a call to *setPushSubscription*. It takes the following argument:
 
-The response to *setPushCallback* is called *pushCallbackSet*. It has the following argument:
+- **pushSubscription**: `PushSubscription|null`
+  The PushSubscription object representing the endpoint the JMAP server will POST events to. This will replace any previously set subscription. Set to `null` to remove any previously registered subscription.
 
-- **callback**: `String|null`
-  Echoed back from the call.
+The response to *setPushSubscription* is called *pushSubscriptionSet*. It has no arguments.
 
-The following error may be returned instead of the *mailboxesSet* response:
+The following errors may be returned instead of the *pushSubscriptionSet* response:
 
 `invalidUrl`: Returned if the URL does not begin with `https://`, or is otherwise syntactically invalid or does not resolve.
 
 `forbidden`: Returned if the URL is valid, but for policy reasons the server is not willing to connect to it.
 
-### getPushCallback
+### getPushSubscription
+
+To check the currently set push subscription (if any), make a call to *getPushSubscription*. It does not take any arguments. The response is called *pushSubscription* and it has a single argument:
+
+- **pushSubscription**: `PushSubscription|null`
+  The PushSubscription object the JMAP server is currently posting push events to, or `null` if none.
+
+## Event Source
+
+Clients that can hold open TCP connections can connect directly to the JMAP server to receive push notifications via a `text/event-stream` resource, as described in <http://www.w3.org/TR/eventsource/>. This is a long running HTTP request down which the server can push data.
+
+When a change occurs in the data on the server, it pushes an event called **state** to any connected clients, with the *StateChange* object as the data.
+
+The server SHOULD also set a new `Last-Event-Id` that encodes the entire server state visible to the user. When a new connection is made to the event-source endpoint, the server can then work out whether the client has missed some changes which it should send immediately.
+
+The server MUST also send an event called **ping** with an empty object as the data if a maximum of 5 minutes has elapsed since the previous event. This MUST NOT set a new `Last-Event-Id`.
```

Comment:
5 minutes is pretty arbitrary.  Why that specific number?  Can I use my knowledge about the network path to choose a longer (and therefore more efficient) value?

Reply:
Yeh, it's arbitrary. The trouble is the client needs to know what value the server is using to know if it should have seen a ping by now! The client could set this in a header when it connects to the server though to request a particular ping time; that's probably better. The main downside is that you can't set custom headers if you are using the browsers' native EventSource object (web clients can still use XHR to handle it though).